### ISSUE TITLE
Cannot open the read only Realm. 'Permission' is missing 

### ISSUE THREAD

**Issue by jaisontj** [realm/realm-java#5414]
ISSUE TYPE: BUG
=============

#### Goal

> Fetch permissions for a user

#### Expected Results

> Successful fetch of permissions

#### Actual Results

>  REALM_JAVA: io.realm.exceptions.RealmMigrationNeededException: Cannot open the read only Realm. 'Permission' is missing.
                                                                   at io.realm.Realm.<init>(Realm.java:165)
                                                                   at io.realm.Realm.createInstance(Realm.java:417)
                                                                   at io.realm.RealmCache.doCreateRealmOrGetFromCache(RealmCache.java:348)
                                                                   at io.realm.RealmCache.createRealmOrGetFromCache(RealmCache.java:281)
                                                                   at io.realm.RealmCache$CreateRealmRunnable.run(RealmCache.java:115)
                                                                   at io.realm.internal.async.BgPriorityRunnable.run(BgPriorityRunnable.java:34)
                                                                   at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:457)
                                                                   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
                                                                   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
                                                                   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
                                                                   at java.lang.Thread.run(Thread.java:764)
                                                               `CreateRealmRunnable` failed.

#### Steps & Code to Reproduce

> Following instructions [here](https://realm.io/docs/java/3.7.1/#retrieving-permissions)

#### Code Sample

```java
public static void fetchPermissions() {
        final PermissionManager pm = SyncUser.currentUser().getPermissionManager();
        // Retrieve
        pm.getPermissions(new PermissionManager.PermissionsCallback() {
            @Override
            public void onSuccess(RealmResults<Permission> permissions) {

                Log.i(TAG, "PRINTING ALL PERMISSIONS");
                printPermissions(permissions);

                // Change listeners can be registered
                // The PermissionManager keeps a reference to these as long as the Permission Manager
                // is open, so they do not risk getting GC'ed
                permissions.addChangeListener(new RealmChangeListener<RealmResults<Permission>>() {
                    @Override
                    public void onChange(RealmResults<Permission> permissions) {
                        Log.i(TAG, "CHANGED PERMISSIONS");
                        printPermissions(permissions);
                    }
                });
                pm.close();
            }

            @Override
            public void onError(ObjectServerError error) {
                // handle error
                Log.i(TAG, "Error fetching permissions");
                Log.i(TAG, error.toString());
                pm.close();
            }
        });
}
```

#### Version of Realm and tooling
Realm version(s): 3.7.1

Realm sync feature enabled: yes

Android Studio version: 2.3.3

Which Android version and device: ?

**Comment by Zhuinden** [comment]
What is your RealmConfiguration?

**Comment by jaisontj** [comment]
I am using Realm Sync, 

``` java
SyncCredentials credentials = SyncCredentials.usernamePassword(ADMIN_CREDENTIALS.username, ADMIN_CREDENTIALS.password, false);
            SyncUser.loginAsync(credentials, AUTH_URL, new SyncUser.Callback() {
                @Override
                public void onSuccess(SyncUser user) {
                    initialiseRealms(user, listener);
                }

                @Override
                public void onError(ObjectServerError error) {
                    listener.failure(error);
                }
            });
```

**Comment by jaisontj** [comment]
I also get another message in the debug log from REALM_JAVA
>REALM_JAVA: Error happened in PermissionManager for 3e9dc2asdasdadwq4564554eeaf925b: UNKNOWN(-1)
                                                               Error occurred in Realm: 
                                                               Default Permission Realm
                                                               UNKNOWN(-1)
                                                               io.realm.exceptions.RealmMigrationNeededException: Cannot open the read only Realm. 'Permission' is missing.
                                                                   at io.realm.Realm.<init>(Realm.java:165)
                                                                   at io.realm.Realm.createInstance(Realm.java:417)
                                                                   at io.realm.RealmCache.doCreateRealmOrGetFromCache(RealmCache.java:348)
                                                                   at io.realm.RealmCache.createRealmOrGetFromCache(RealmCache.java:281)
                                                                   at io.realm.RealmCache$CreateRealmRunnable.run(RealmCache.java:115)
                                                                   at io.realm.internal.async.BgPriorityRunnable.run(BgPriorityRunnable.java:34)
                                                                   at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:422)
                                                                   at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
                                                                   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
                                                                   at java.lang.Thread.run(Thread.java:818)

**Comment by jaisontj** [comment]
Hi @Zhuinden Any update on this ? Just a pointer to the right direction to look into will help.

**Comment by beeender** [comment]
@jaisontj what ROS version are you using?

**Comment by jaisontj** [comment]
@beeender I'm using ROS version v1.8.3

**Comment by beeender** [comment]
@jaisontj made a fix in #5418 

There is a race condition caused it. Before it gets released, you can try to clear the app data and restart the app.

**Comment by jaisontj** [comment]
@beeender I have already done that multiple times. When will the fix be released?
